Add `SystemInfo` utility class which provides the system information

trustin opened the pull request:
Motivation:

It'd be nice for Armeria to provide a reliable way to retrieve the
system information that requires some hack depending on OS and Java
version, such as:

- Java version
- local hostname
- current PID
- microsecond-precision time

Modifications:

- Added `SystemInfo` which provides the information mentioned in the
  'Motivation' section.

Result:

- Convenience.

trustin commented:
@anuraaga IIRC you requested a utility class for getting the local hostname long time ago, so it'd be nice if you have a suggestion for a better class name, etc.

trustin commented:
A reminder to self: On Linux, you can check the symlink `/proc/self`:

```
$ ls -l /proc/self
lrwxrwxrwx 1 root root 0 Mar 12 09:37 /proc/self -> 7628
```

anuraaga commented on core/src/main/java/com/linecorp/armeria/common/util/MoreSystem.java:
```diff
@@ -0,0 +1,224 @@
+/**
+ * Provides utilities for accessing the information about the current system and process.
+ */
+public final class MoreSystem {
```
I guess this is following the `More` convention with the `System` class. It does read quite strangely though. Do you see it being more than just providing system information? If not, could just be `SystemInformation` or similar.

trustin replied:
Asked the naming question at Twitter and got the following:

- `Os`
- `PosixUtil`
- `LocalServerInfo`
- `SystemHosts`
- `CurrentSystem`
- `LocalSystem`
- `Sys`

Perhaps, `SystemInfo`?

anuraaga replied:
+1 to `SystemInfo`

anuraaga commented on core/src/main/java/com/linecorp/armeria/common/util/MoreSystem.java:
```diff
@@ -0,0 +1,224 @@
+            // Try ProcessHandle.pid() if Java 9+.
+            if (PlatformDependent.javaVersion() >= 9) {
+                try {
+                    final Class<?> handleClass = Class.forName("java.lang.ProcessHandle", true,
```
What do you think about using class loading instead of reflection? I guess it might complicate the build though

trustin replied:
Yeah, `ProcessHandle` is not available in Java 8 so it will complicate the build, especially that we pass `--release 8` to the compiler universally. Let me keep this because it simplifies the build. :-)

trustin commented:
- Renamed to `SystemInfo`
- Added `javaVersion()` and `currentTimeMicros()`

I feel like `currentTimeMicros()` might be better hosted somewhere else, but perhaps it's just fine to keep it here.

trustin commented:
Bonus: On Linux, hostname is retrieved from `/proc/sys/kernel/hostname` which means we now retrieves the hostname even on a system without the `hostname` command. I guess it will make @anuraaga happy. :wink:

trustin commented on core/src/main/java/com/linecorp/armeria/common/util/SystemInfo.java:
```diff
@@ -0,0 +1,341 @@
+        static final String HOSTNAME;
+
+        static {
+            // Try /proc/sys/kernel/hostname' on Linux.
```
Fixed :-)

trustin commented:
Thanks, @minwoox @hyangtack !
